core/src/main/java/com/linecorp/armeria/common/stream/DeferredStreamMessage.java

```diff
@@ -169,20 +185,24 @@ private void subscribe0(Subscriber<? super T> subscriber, Executor executor, boo
             failLateSubscriber(subscriber, executor, AbortedStreamException.get());
             return;
         }
-
         final PendingSubscription<? super T> newPendingSubscription =
                 new PendingSubscription<>(subscriber, executor, withPooledObjects);
 
-        final StreamMessage<T> delegate = this.delegate;
-        if (delegate == null) {
-            if (!pendingSubscriptionUpdater.compareAndSet(this, null, newPendingSubscription)) {
-                failLateSubscriber(subscriber, executor,
-                                   new IllegalStateException("subscribed by other subscriber already"));
-            }
+        if (!subscriptionUpdater.compareAndSet(this, null, newPendingSubscription)) {
+            failLateSubscriber(subscriber, executor,
+                               new IllegalStateException("subscribed by other subscriber already"));
             return;
         }
 
-        subscribeToDelegate(delegate, newPendingSubscription);
+        for (;;) {
+            if (stateUpdater.compareAndSet(this, State.NONE, State.SUBSCRIBED)) {
+                return;
+            }
+            if (stateUpdater.compareAndSet(this, State.DELEGATED, State.COMPLETE)) {
```

Comment:
Could you add some comments here about why this seemingly infinite loop will never loop infinitely?